clippy_lints/src/map_nil_fn.rs

```diff
@@ -0,0 +1,187 @@
+use rustc::hir;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::codemap::Span;
+use utils::{in_macro, iter_input_pats, match_type, method_chain_args, snippet, span_lint_and_then};
+use utils::paths;
+
+#[derive(Clone)]
+pub struct Pass;
+
+/// **What it does:** Checks for usage of `Option.map(f)` where f is a nil
+/// function or closure
```

Comment:
Given that this code also handles uninhabited/never types, it's not just unit, so we might want to keep the function names the same. However, the user-facing lint name and documentation should probably just handle unit; using a diverging expression in `map` sounds like a very rare use case.